### ISSUE TITLE
Can't manipulate sharded f64 array on TPU

### ISSUE THREAD

**Issue by tomdbar** [google/jax#15225]
### Issue summary

I can shard f32 and f64 arrays across multiple devices, however, on TPUs when the distributed f64 arrays are used for even simple operations, the output is _replicated_ across all devices, instead of respecting the input sharding.

### Minimum working example

```
# Uncomment below for local CPU testing
#import os
#flags = os.environ.get('XLA_FLAGS', '')
#os.environ['XLA_FLAGS'] = flags + " --xla_force_host_platform_device_count=8"

import jax
import jax.numpy as jnp
import numpy as np
from jax.config import config
from jax.sharding import Mesh, NamedSharding, PartitionSpec

config.update("jax_enable_x64", True)

if __name__=="__main__":
    mesh_devices = np.array(jax.devices()).reshape(4,2)
    mesh = Mesh(mesh_devices, axis_names=('x', 'y'))
    partition_spec = PartitionSpec('x', 'y')
    sharding = NamedSharding(mesh, partition_spec)

    dim = 64
    for dtype in [jnp.float32, jnp.float64]:
        A = jax.device_put(np.ones((dim, dim)), sharding).astype(dtype)

        print(f"Input sharding (dtype={A.dtype})...")
        jax.debug.visualize_array_sharding(A)

        B = 2*A

        print(f"Output sharding (dtype={B.dtype})...")
        jax.debug.visualize_array_sharding(B)
```

Output on a v3-8 TPU:

```
Input sharding (dtype=float32)...
┌──────────┬──────────┐
│  TPU 0   │  TPU 1   │
├──────────┼──────────┤
│  TPU 2   │  TPU 3   │
├──────────┼──────────┤
│  TPU 4   │  TPU 5   │
├──────────┼──────────┤
│  TPU 6   │  TPU 7   │
└──────────┴──────────┘
Output sharding (dtype=float32)...
┌──────────┬──────────┐
│  TPU 0   │  TPU 1   │
├──────────┼──────────┤
│  TPU 2   │  TPU 3   │
├──────────┼──────────┤
│  TPU 4   │  TPU 5   │
├──────────┼──────────┤
│  TPU 6   │  TPU 7   │
└──────────┴──────────┘
Input sharding (dtype=float64)...
┌──────────┬──────────┐
│  TPU 0   │  TPU 1   │
├──────────┼──────────┤
│  TPU 2   │  TPU 3   │
├──────────┼──────────┤
│  TPU 4   │  TPU 5   │
├──────────┼──────────┤
│  TPU 6   │  TPU 7   │
└──────────┴──────────┘
Output sharding (dtype=float64)...
┌───────────────────────┐
│                       │
│                       │
│                       │
│                       │
│  TPU 0,1,2,3,4,5,6,7  │
│                       │
│                       │
│                       │
│                       │
└───────────────────────┘
```

Output on CPU's (which behave as I would expect/desire):

```
Input sharding (dtype=float32)...
┌──────────┬──────────┐
│  CPU 0   │  CPU 1   │
├──────────┼──────────┤
│  CPU 2   │  CPU 3   │
├──────────┼──────────┤
│  CPU 4   │  CPU 5   │
├──────────┼──────────┤
│  CPU 6   │  CPU 7   │
└──────────┴──────────┘
Output sharding (dtype=float32)...
┌──────────┬──────────┐
│  CPU 0   │  CPU 1   │
├──────────┼──────────┤
│  CPU 2   │  CPU 3   │
├──────────┼──────────┤
│  CPU 4   │  CPU 5   │
├──────────┼──────────┤
│  CPU 6   │  CPU 7   │
└──────────┴──────────┘
Input sharding (dtype=float64)...
┌──────────┬──────────┐
│  CPU 0   │  CPU 1   │
├──────────┼──────────┤
│  CPU 2   │  CPU 3   │
├──────────┼──────────┤
│  CPU 4   │  CPU 5   │
├──────────┼──────────┤
│  CPU 6   │  CPU 7   │
└──────────┴──────────┘
Output sharding (dtype=float64)...
┌──────────┬──────────┐
│  CPU 0   │  CPU 1   │
├──────────┼──────────┤
│  CPU 2   │  CPU 3   │
├──────────┼──────────┤
│  CPU 4   │  CPU 5   │
├──────────┼──────────┤
│  CPU 6   │  CPU 7   │
└──────────┴──────────┘
```

### What jax/jaxlib version are you using?

jax v0.4.4 jaxlib v0.4.4

### Which accelerator(s) are you using?

TPU

**Comment by skye** [comment]
Thanks for the report!

Internal bug for Googler's reference: b/275403738

**Comment by skye** [comment]
I believe this should be fixed via https://github.com/openxla/xla/commit/6586b8b816ba2054f90fc3824a33264a5ae132b7 (thanks @tongfei-guo!). It'll be included in the next jax+jaxlib release (0.4.10, hopefully will be this week). I'm gonna close this issue lest I forget later, but please reopen if you find it's still an issue with the next release.
